### DISCUSSION TITLE
Managing slow jobs and servers

### DISCUSSION THREAD

**Discussion by reallyuniquename** [sabnzbd/sabnzbd#2176] [Q&A]
Say, I have this setup.
```
Server A: fast, backbone A, priority 1.
Server B: slow, backbone B, priority 2.
```
Job1 is only available on server B. Job2, job3 and job4 are available at 99% on server A and at 1% on server B.

If I set top of the queue like this
```
Job1
Job2
Job3
Job4
...
```
Then I get crawling job1 and job2, job3 and job4 eventually stuck at 99% waiting for job1 to finish. They still need to check that 1% of missing articles against server B.

If I set top of the queue like this
```
Job2
Job3
Job4
Job1
...
```
Then job1 waits for job2, job3 and job4 to finish with server A. And that never happens untill those jobs are completed because server A is always busy. Server B will be idle most of the time.

Now I want job1 to constantly pull articles from server B while also allowing other jobs to occasionally occupy server B for a short time. Is there a setup that makes that possible?

Setting servers to the same priority wouldn't work well for block accounts.

**Comment by puzzledsab** [comment]
The queue system may not work the way you would think intuitively. If you let server A check all the articles of Job1 and then move it to the bottom of the queue then Server B will continue downloading articles from it. If Job2-Job4 have missing articles then Server B will fetch them and then continue on Job1. Make sure "Only Get Articles for Top of Queue" is unchecked.

There's no way to automate what you're asking for, though.

**Reply by reallyuniquename** [reply]
> If you let server A check all the articles of Job1 and then move it to the bottom of the queue then Server B will continue downloading articles from it.

Yeah, I already do that, the queue system is actually quite intuitive. But that requires manual control and I'm trying to figure if this behaviour could be automated. Maybe some kind of option to drop slow jobs to the bottom of the queue?

**Reply by sanderjo** [reply]
> But that requires manual control 

Based on what? Low speed of a certain download? What is a typical high speed download for you? And a slow speed download?

> and I'm trying to figure if this behaviour could be automated. 

I think you can, with a script talking to the SABnzbd API:

1. From API-History get the typical high speeds. Then set (for example) high-speed to 70% the value of highest speed
2. The, from API-Queue, get the current speed (if downloading).
3. If downloading: if the current speed is below 70%-of-highest-speed, set the current queue-item to prio Low; apparently it's slow

... and let that check run each 1 minute (from crontab for example, or just loop with a 1 minute sleep)

On which OS is your SAB running?

> Maybe some kind of option to drop slow jobs to the bottom of the queue?

If the above method works and is useful, maybe it can be put into SAB (if @Safihre approves)

**Reply by albino1** [reply]
Even if my downloads start at 100MB/s they slow down significantly in the last few percent, so if you re-arranged queue based on some sort of speed metric then every download would get immediately dropped to the bottom of the queue just as it was about to finally finish.

**Reply by sanderjo** [reply]
> Even if my downloads start at 100MB/s they slow down significantly in the last few percent, so if you re-arranged queue based on some sort of speed metric then every download would get immediately dropped to the bottom of the queue just as it was about to finally finish.

If so, solved by not changing order if in last 10% of download ... ?

**Reply by albino1** [reply]
If you had a huge queue and kept changing the order by shifting everything to the bottom when it was at like 99% then seemingly nothing would ever finish.  So yeah, you're right, you would have to build in all sorts of logic, like if > 95% complete then don't re-order, etc.  Seems like a lot of effort for a very niche problem, though I'll admit that this would occasionally be useful for me as well.

Regarding the OP, I would just change your priorities to make them all 0 if that's a concern.

**Reply by puzzledsab** [reply]
An alternative could be to do a check every minute or so to move stalled nzbs (0 bps download speed) with less than 98% or whatever remaining to the top. In most cases those would probably finish in a few seconds but maybe skip rearranging if the top nzb already has less than 98% remaining.

**Reply by reallyuniquename** [reply]
I guess I described that badly. It's not always about speed. I'm talking about slow jobs as SABnzbd defines them. Basically a job is slow when servers can't keep up with each other while requesting job's articles. So you quickly get idling server A and full queue for server B.

There is already an option called `Only Get Articles for Top of Queue. Disable to prevent slow jobs from blocking the queue` that almost solves this.

So, instead of just utilizing server A for the next job I suggest to also lower the priority for job1. Shouldn't that work? Am I missing something here?

**Reply by puzzledsab** [reply]
To try to make server B keep up you can set as few connections as possible for server A and almost maximum for server B. That way B will take up more of the available bandwidth when it is in use. If B takes much longer to check and download the articles than A then there no way to stop A from idling at the end because it will finish first.

In those cases it can be helpful to let almost finished jobs get top priority, which means moving them to the top of the queue. The average download time will be the same but the mean time will go down. That's because the nzbs that take a long time would take a little bit longer but the almost finished ones would be finished much earlier.

I don't know what you mean by slow jobs. I don't think SABnzbd defines anything as a slow job. Trying to identify and change the priority of jobs that hold other jobs back sounds much more complex than moving almost finished jobs to the top of the queue. "Only Get Articles for Top of Queue" wouldn't help. In most cases it would probably be even slower because A has to wait for the slower B after every NZB.

@Safihre : What's your opinion about this? I can try implementing an option for it if you think it's worth it. I've thought about it before because having a lot of almost finished jobs fills up the temp partition. I usually solve it by manually moving them to the top but it's a hassle and the temp disk may have been full for a while before I notice. My idea is to add an option for the minimum percentage of articles that needs to be downloaded before a job with 0 BPS download speed gets moved to the top.

**Reply by Safihre** [reply]
@puzzledsab We could add another "Sort queue" option that sorts by completion? Then that could be called every time a new job hits the queue. We could then for example not adjust the order for all jobs with <20% completion to not create a random order for all jobs that were not really started yet,
Indeed it sounds a bit too exotic for me to do some auto-detection.
What do you think?

**Reply by reallyuniquename** [reply]
>I don't know what you mean by slow jobs. I don't think SABnzbd defines anything as a slow job.

Then how does unticking "Only Get Articles for Top of Queue. Disable to prevent *slow jobs* from blocking the queue" work? Is it not aware of jobs? 

Surely there has to be a moment "we are stalling with job1 let's fire up job2". That's when you can tag current job1 as a slow one. Basically if job1 is not finished and job2 is already being downloaded then job1 is "slow". Push job1 all the way down, set its priority to `Low`. Let it pull articles from the slow server but don't block other jobs from doing that occasionally. Does that make sense? I do it manually and SABnzbd handles it just fine.

That logic should not require any hardcoded values like speed or completion otherwise implementing it is probably not worth your time.

**Reply by Safihre** [reply]
All that setting does is prevent ever picking up job2. The explanation is just there to warn that if job1 is slow, it will block everything.

**Reply by reallyuniquename** [reply]
Then there has to be some kind of handler that puts job2 into the active queue while job1 is still running, right?

**Reply by puzzledsab** [reply]
@reallyuniquename All nzbs that are not paused are in the "active" queue all the time. A bit simplified, when a server has a free connection it will start looping through the queue to find an article that hasn't been tried by that server. If it has tried all the articles from the first nzb then it starts checking the next.

However, if "Only Get Articles for Top of Queue" is checked then it will stop looking for articles after the first nzb. All the servers will have to have checked all the articles of the first nzb before any of them are allowed to go on to the next. The system has no awareness of the download speed of each server. The reason why the top server is way ahead may be that it doesn't have most of the articles and returns not found very quickly.

@Safihre The use case I'm thinking of is adding 100 nzbs in a batch, where some will reach 99% and some 50% with the top pri server. There might be 50 or more unfinished jobs in the queue after a while. No more nzbs are being added, so rearranging the queue would need to be triggered by something other than adding new nzbs.

Rearranging all started nzbs by finished status regularly might be ideal in some cases, but it could also mess more with the priorities defined by the user. I think it would also require noticeably more complex code than moving an nzb that has reached x% to the top but I haven't actually tried. My thinking was to get rid of the worst case scenarios but maybe rearranging all started nzbs regularly would be better for most users.

**Reply by thezoggy** [reply]
rather than doing the whole re-arranging bit (which imho is prob not ideal for most) why not just have a counter when a current job has had all the articles checked/retried, then you just basically set that job to pause (to crudely marked its slow, or use that counter flag to know) and move onto next job. rinse and repeat. on the following job completion (or marked as slow) you reset the iteration to start the jobs over. let the first job try again, if it still slow you mark and move on. creates a waterfall approach where the top jobs get re-tested over time (because might just have propagation issues / need to be retried once connections timed out) but more or less still hits up the next jobs with minimal impact to previous. and then majority of the time, this logic wont really get used because most jobs arent 'slow'

**Reply by thezoggy** [reply]
but anyways, if a job cant be completed it should have just bailed out. if you seem to have this issue a lot it makes me wonder why you dont just have completion check enabled or not set to proper amount? also how many servers/ has backup server?

**Reply by reallyuniquename** [reply]
@puzzledsab I get how queue works. I mean that job is active when its articles are actually being tried by any server.

>if "Only Get Articles for Top of Queue" is checked then it will stop looking for articles after the first nzb

But I have it unchecked to speed things up. There is a moment when job1 is being actively downloaded and SABnzbd simultaneously starts downloading job2. This is it, job1 is "slow". Set job1 priority to low. There is no need for awareness of the download speed or completion. I do that manually without any specific numbers in my mind and I'm having trouble understanding why this can't be easily automated. What am I missing here?

**Reply by reallyuniquename** [reply]
@thezoggy
This is not about completion. All jobs in my example can be completed using given servers. The only issue here is effective use of time. If job1 takes 24 hours to complete then every other job with at least one missing article will take roughly additional 24 hours to complete. If I initially send job1 to the bottom of the queue then job1 will take all the time it takes download other jobs plus 24 hours which is not ideal for huge batches either. You see the problem?


>why you dont just have completion check enabled or not set to proper amount?

Would it help? Would SABnzbd first check every job availability and only then start pulling articles? I also don't how that option works so I fear I can hit false negatives.

**Reply by thezoggy** [reply]
it is about completion, its just that x server may not be able to complete it on its own. maybe missing articles and had to pull from backup/block or just simply retry the article again later on the same server.

**Reply by puzzledsab** [reply]
@reallyuniquename You can see from the GUI which nzbs are currently being downloaded but SABnzbd doesn't actually have a concept of active jobs. It just registers when data has been downloaded for an nzb and updates the bandwidth usage. This is not important for this issue, though.

The fact that server A skips to job 2 in the queue doesn't mean that job 1 is slow, it only means that Server A couldn't provide all the articles. Server B might finish Job 1 in a second or it might take days. If it is automatically given a lower priority because Server A skipped ahead then Server B will start downloading missing articles for job 2 instead, even if job 1 is at 99%. The odds that you end up with a bunch of jobs that are 99% finished and waiting for a job that's missing a lot of articles would be the same, even though it would happen under slightly different circumstances.

**Reply by Safihre** [reply]
Indeed.

**Reply by reallyuniquename** [reply]
@puzzledsab Now I'm really confused. If completion makes a difference here and I don't see how, assume that all NZBs in the queue can be completed with a given setup.

>The fact that server A skips to job 2 in the queue doesn't mean that job 1 is slow, it only means that Server A couldn't provide all the articles

If articles are not found on fast server A then they are going to be downloaded from slow server B. How does that not make job1 slow?

>If it is automatically given a lower priority because Server A skipped ahead then Server B will start downloading missing articles for job 2 instead, even if job 1 is at 99%.

How would job1 that is only available on slow server B get ahead of job2 available on server A? Server A is blazing fast and it has the highest priority. When it skips ahead it happens very early.

>The odds that you end up with a bunch of jobs that are 99% finished and waiting for a job that's missing a lot of articles would be the same

No? Why would higher priority jobs wait for lower priority jobs? At least that's not what my instance of SABnzbd does. The moment job1 gets lower priority missing articles from those jobs-at-99% start seeping into server B queue even though it's still busy with job1. I could literallly see this on `Connections` tab.

**Reply by Safihre** [reply]
I can see how the proposal of @reallyuniquename would work. I'm still not sure if this is useful for a lot of regular SABnzbd users. How often does one have such significant speed differences?
Personally, I don't think this is very common.

**Reply by puzzledsab** [reply]
I agree that it would work as intended if there is only one or very few missing articles but sorting by completed status would work just as well and was much easier to implement. See #2186.

@reallyuniquename : All your jobs have missing articles, that's why they stop at 99%. If the missing articles are at the end of the download then there wouldn't need to be many before they all risk getting low priority. I think my implementation solves both our needs.

**Reply by reallyuniquename** [reply]
So slow jobs will be slowly pushed down as downloads go, right? Actually that's probably even better. Although how are you going to trigger resort? Is there going to be some configurable timer? My only concern is that download order for other 0% jobs can be completely mangled.

**Reply by albino1** [reply]
Looking over the PR, I wonder if you intentionally tried to manually move something to the top if it would just get moved back down again by the % sort. 

For example, if you had a 50gb download that was at 20%, and then you added a small 1gb download below it and hit the Top up arrow button, then it would just get auto-sorted back down under the 50gb file over and over.

I do think this is really interesting though, and something that I would probably enable, but I would get annoyed if I moved something to the top to prioritize it only for it to get de-prioritized automatically.

**Reply by thezoggy** [reply]
ideally it should still be a weighted queue with stable sort.

i'd expect priority still to win, then everything else sorted by completed%.

so if you have 10 downloads and one is forced, forced should be above the rest with the other 9 sorted by comp%.
i could see an issue when several downloads are stuck at the same % and maybe they are shuffling around if the order is not stable.

**Reply by albino1** [reply]
"Priority" was a poor choice of terms on my part, as in my head I was explicitly excluding Forced/High/etc, which are literally listed under `Priority`.  I was thinking more just manual ordering where you move things up and down based on what you think is most important in that moment.  You're right in that if you set it to Force or High then that would probably fix it, and would take precedence over % sort.  You kind of lose the ability to manually sort multiple items in terms of which order you would prefer, but if that's that big of a deal then I suppose you shouldn't be using % sort then.

**Reply by puzzledsab** [reply]
@reallyuniquename Yes, as soon as another job has more percent of data downloaded. Currently it is resorted automatically every 30 seconds if this sorting method is selected in the config. The order of untouched jobs will not be changed by the resorting. You can download installable versions for Windows or Mac here if you want to try it: https://github.com/sabnzbd/sabnzbd/actions/runs/2367698020

@albino1 If you move it to the top with no priority change then it will download for up to 30 seconds until the list is resorted and then probably be moved back down a bit, but above all the files with 0 bytes downloaded. The nzbs are only resorted between their own priority level so if you give it a higher priority than the other ones then it will always be on top.

**Reply by reallyuniquename** [reply]
What's the worst case scenario here? Say, both server A and server B speeds start fluctuating. Personally, I don't think I've ever experienced that. Just a bunch of jobs, probably only two, fighting over #1 place in the queue every 30 seconds, right?

>resorted automatically every 30 seconds

For a moment I was going to ask if this value will be tweakable somewhere but then I realized that 30 seconds timer is tighly connected to other tasks and variables.

**Reply by puzzledsab** [reply]
I doubt there will be much fluctuating. Maybe for a short while in special cases but I think that normally the job that caught up with the other one should benefit additionally from being on top. Having multiple backup servers with different priority levels would probably increase the chances of fluctuation somewhat.

**Comment by ErikBrown2** [comment]
Sorry to butt in here. After reading the above I am confused about a slightly related question. I have a main server with priority 0 and a backup server with priority 1 configured. The only purpose of the backup server is to provide articles that are missing in the main server. I do not care about speed issues. I understand from the above that this will not work automatically or do I misunderstand this?

**Reply by puzzledsab** [reply]
You misunderstand, that part works. This change is mainly for special cases where for instance a job that could be finished in a second will have to wait a long time before the backup server gets to it because it's busy downloading a job that is missing much more data.
